src/core/utils/device_qualifier.hpp

```diff
@@ -0,0 +1,16 @@
+#ifndef UTILS_DEVICE_QUALIFIER_HPP
+#define UTILS_DEVICE_QUALIFIER_HPP
+
+#if defined(__CUDACC__)
+#define DEVICE_QUALIFIER __host__ __device__
+#else
+#define DEVICE_QUALIFIER
+#endif
+
+#ifdef __CUDACC__
+#define DEVICE_ASSERT(A) assert((A))
```

Comment:
Why not have all defines in on ```ifdef``` ?

Reply:
Not the same logic

Reply:
it's binary logic that depends on ```__CUDACC__```, and both cases are handled at the top. What do I miss here?

Reply:
one is `ifndef`